### PR TITLE
refactor(module/elemwise): use enum instead of method string

### PR DESCRIPTION
**Pull request by bigeagle**
use python `Enum` instead of `str` to indicate elemwise operation.

### CODE REVIEW

**Review by xxr3376 — COMMENTED** [review]
Yes, we need to report KeyError when a user creating a module with the wrong method name. But it's really awful that we need to type 3 times when adding a new method. 

We do have the enum underlying, e.g. `megengine.functional.elemwise.Elemwise.Mode.LOG`, but it doesn't provide a user-friendly interface (can't initialize by calling `Mode('ABS')`).

I think there are 3 ways to make things better.

1. Use underlying one directly (It's a little bit implementation-related, may not suitable to expose to end-user)
2. Keep this module's constructor accept `str` only, but check it by calling `getattr`
3. Module use a separated enum class, and generate this enum class by reading underlying one

**Comment by bigeagle** [comment]
> 1. Use underlying one directly (It's a little bit implementation-related, may not suitable to expose to end-user)
> 2. Keep this module's constructor accept str only, but check it by calling `getattr`
> 3. Module use a separated enum class, and generate this enum class by reading underlying one

Using enum has an advantage that editor's auto-completion can suggest available methods.
Method 1 may not be suitable to expose to end-user, we should wrap it as an enum anyway, or use code-gen like method3.
I would prefer method 2 if we can utilize `Literal` type hinting from [PEP586](https://www.python.org/dev/peps/pep-0586/)

**Comment by xxr3376** [comment]
PEP586 seems promising in the future, I'd like to keep this API accept `str` only to keep API as simple as possible.
I saw two drawbacks about PEP586 + method2:

1. Currently, we release MegEngine for python `3.5`, `3.6`, `3.7,` `3.8`. Neither PEP586 nor statements like `self.method: ElemwiseMethod = method` can be used here. -> Maybe we can write some simple code, just provide auto-completion when python>=3.8?
2. User may difficult to iterate all methods like an enum


I'm not sure whether all editors still provide good completion when we set the type of `method` to `Union[ElemwiseMethod, str]` (to keep API compatible)? Would you please check it in your editor?


P.S. Currently, usages are pretty different between `functional.elemwise` and `module.Elemwise`, do you think it's a good idea to make this API accept functional method? User's code may look like `module.Elemwise(F.elmwise.add)`.

**Comment by fzhangscorec** [comment]
The changes look good to me.  I can see the enum class serves two purpose: make users's coding easier and help check API usage error.  

@bigeagle maybe you want to change the commit message. The api accept both ''str" (the old way, also for backward compatibility) and new enum.

**Comment by bigeagle** [comment]
> 1. Currently, we release MegEngine for python `3.5`, `3.6`, `3.7,` `3.8`. Neither PEP586 nor statements like `self.method: ElemwiseMethod = method` can be used here. -> Maybe we can write some simple code, just provide auto-completion when python>=3.8?

[typing-extentions](https://pypi.org/project/typing-extensions/) has backported PEP586 to python3.5+

> 2. User may difficult to iterate all methods like an enum
[typing.get_args()](https://stackoverflow.com/questions/59163553/getting-the-literal-out-of-a-python-literal-type-at-runtime) can implement this.

> I'm not sure whether all editors still provide good completion when we set the type of `method` to `Union[ElemwiseMethod, str]` (to keep API compatible)? Would you please check it in your editor?

My editor can only provide auto completion when I type `ElemwiseMethod.`

> P.S. Currently, usages are pretty different between `functional.elemwise` and `module.Elemwise`, do you think it's a good idea to make this API accept functional method? User's code may look like `module.Elemwise(F.elemwise.add)`.

I don't quite agree with this API, it reminds me of something like `module.ModuleWrapper(F.relu)`.

**Comment by xxr3376** [comment]
> I don't quite agree with this API, it reminds me of something like `module.ModuleWrapper(F.relu)`
I think there is a little bit difference

1. `module.Elemwise(module.ElemwiseMethod.Add)` and `module.Elemwise("ADD")` and `module.Elemwise(F.elemwise.add)` are same wrapper function, I don't think there's any difference between them when comparing with 
`module.ModuleWrapper(F.relu)`.
2. We only accept `F.elemwise`, this is a more strict constraint. 

Here's my personal flavor about these usages(1 > 2 > 3 > 4) , what about your opinion?

1. `module.elemwise.Add`
2. `module.Elemwise(F.elemwise.add)`
3. `module.Elemwise("ADD")`
4. `module.Elemwise(module.ElemwiseMethod.Add)`

Here's the principle about this choice:

* Usages between `functional.elemwise` and `module.Elemwise`: more consistent is better
* Less entity is better
* shorter is better

**Comment by bigeagle** [comment]
> Here's my personal flavor about these usages(1 > 2 > 3 > 4) , what about your opinion?
> 1. `module.elemwise.Add`
> 2. `module.Elemwise(F.elemwise.add)`
> 3. `module.Elemwise("ADD")`
> 4. `module.Elemwise(module.ElemwiseMethod.Add)`

IMO, 1 > 2 > 4 > 3 , simply because I dislike schema-less enumeration.
For case 1, there will be confusing APIs like `module.activation.ReLU` and `module.elemwise.ReLU`.
I think `2` is overall a better choice, and the function signature can be designed as
```python
class Elemwise(Module):
    def __init__(self, method: ElemwiseFunction):
          ...
```

> Here's the principle about this choice:
> 
> * Usages between `functional.elemwise` and `module.Elemwise`: more consistent is better
> * Less entity is better
> * shorter is better

I totally agree.

**Comment by lixiangyin666** [comment]
其实目前Elemwise mode的enum已经内置支持了，用户可以参考[这里](https://github.com/MegEngine/MegEngine/blob/release-1.5/imperative/python/megengine/functional/elemwise.py#L14)import Elemwise，然后可以像[这样](https://github.com/MegEngine/MegEngine/blob/release-1.5/imperative/python/megengine/functional/elemwise.py#L116)用enum了，并且内置了str到enum的隐式转换。
